---
title: "DAY2 GSE198256"
author: "David Gomez-Cabrero"
date: "2024-January"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Experimental design

Lets review experimental design from a practical perspective

```{r read data}

# Read data
urld <- "https://www.ncbi.nlm.nih.gov/geo/download/?format=file&type=rnaseq_counts"
path <- paste(urld, "acc=GSE198256", "file=GSE198256_raw_counts_GRCh38.p13_NCBI.tsv.gz", sep="&");
GSE198256_count <- as.matrix(data.table::fread(path, header=T, colClasses="integer"), rownames=1)

# Read Meta data
library(GEOquery)
gds <- getGEO("GSE198256")
Meta_GSE198256 <- pData(gds$GSE198256_series_matrix.txt.gz@phenoData)
Group <- Meta_GSE198256[,c("disease state:ch1")]

dim(GSE198256_count)
Group

```

## Limma: Normalize and set design

```{r Normalize and set design}

# set DGE class
require(limma)
require(edgeR)
dge <- DGEList(counts=GSE198256_count)

# Make sure on the metadata
rownames(Meta_GSE198256)==colnames(GSE198256_count)
Group[Group=="Covid19: Acute infection"] <- "Covid19AI"
Group[Group=="Covid19: Recovery 3Mo"] <- "Covid193Mo"
Group[Group=="Covid19: Recovery 6Mo"] <- "Covid196Mo"

Group <- factor(Group)
contrasts(Group) <- contr.treatment(levels(Group), base = which(levels(Group) == 'Healthy')) # to make healthy as the intercept 
design <- model.matrix(~ Group )

# Filter
keep <- filterByExpr(dge, design=design)
dge <- dge[keep,,keep.lib.sizes=FALSE]

# Normalization
dge <- calcNormFactors(dge)


```

## Limma: Voom or Trend?

```{r Voom or Trend}

## Trend

# If the sequencing depth is reasonably consistent across the RNA samples, then the simplest and most robust approach to differential exis to use limma-trend. This approach will usually work well if the ratio of the largest library size to the smallest is not more than about 3-fold.
logCPM <- cpm(dge, log=TRUE, prior.count=3)
 # The prior count is used here to damp down the variances of logarithms of low counts.
lmfit <- lmFit(logCPM, design)

lmfit <- eBayes(lmfit, trend=TRUE)
 # logical, should an intensity-dependent trend be allowed for the prior variance? If FALSE then the prior variance is constant. Alternatively, trend can be a row-wise numeric vector, which will be used as the covariate for the prior variance.
 # The use of eBayes or treat with trend=TRUE is known as the limma-trend method (Law et al, 2014; Phipson et al, 2016). With this option, an intensity-dependent trend is fitted to the prior variances s2.prior

topTable(lmfit, coef=ncol(design))


## Voom

# When the library sizes are quite variable between samples, then the voom approach is theoretically more powerful than limma-trend.
v <- voom(dge, design, plot=TRUE)
# The voom method is similar in purpose to the limma-trend method, which uses eBayes or treat with trend=TRUE. The voom method incorporates the mean-variance trend into the precision weights, whereas limma-trend incorporates the trend into the empirical Bayes moderation. The voom method takes into account the sequencing depths (library sizes) of the individual columns of counts and applies the mean-variance trend on an individual observation basis. limma-trend, on the other hand, assumes that the library sizes are not wildly different and applies the mean-variance trend on a genewise basis. As noted by Law et al (2014), voom should be more powerful than limma-trend if the library sizes are very different but, otherwise, the two methods should give similar results.
vfit <- lmFit(v, design)
vfit <- eBayes(vfit)
topTable(vfit, coef=ncol(design))


```

## ACTIVITY 1:

-   How would you compare the results between voom and trend?
-   Is it required to run more than analysis?
-   What exactly are we asking with this differential expression?

```{r ACTIVITY 1}
### comparing the results of voom vs trend 
lmfit_res <- lmfit$coefficients
vfit_res <- vfit$coefficients

### comparing the results of voom vs trend using correlation 
lmfit_res <- topTable(lmfit, coef=ncol(design), number = 50, sort.by = "none")
vfit_res <- topTable(vfit, coef=ncol(design), number = 50, sort.by = "none")
vfit_res <- vfit_res[order(match(rownames(vfit_res),rownames(lmfit_res))), ]
rownames(lmfit_res) == rownames(vfit_res) # making sure that the order is matched for fair comparison 

corr <- cor(lmfit_res$logFC, vfit_res$logFC)
plot(vfit_res$logFC, lmfit_res$logFC, 
     xlab = "Voom Log-Fold Change", ylab = "Trend Log-Fold Change",
     main = "Comparison of Limma-Trend and Voom Results")


```

## ACTIVITY 2:

-   Plan the next analysis: questions, steps,...

```{r ACTIVITY 2}

# To visualize our design: design matrix <- model.matrix(+ Group )
library(pheatmap)
design
pheatmap(design, cluster_rows = F, cluster_cols = F)

# We will be using voom for fitting our model 
vfit
coef.vfit <- vfit$coefficients
head(coef.vfit)

# PACKAGES
#BiocManager::install("clusterProfiler", update = FALSE)
#BiocManager::install("ggupset", update = FALSE)
#BiocManager::install("msigdbr", update = FALSE)
#BiocManager::install("org.Hs.eg.db", update = FALSE)

library(clusterProfiler)
library(msigdbr)
library(org.Hs.eg.db)
library(magrittr)

# Add more contrasts
v <- voom(dge, design, plot=TRUE)
colnames(design) <- c("Intercept","Covid193Mo","Covid196Mo","Covid19AI")
fit <- lmFit(v, design)

contrast.matrix <- makeContrasts(Covid19AI-Covid193Mo, Covid193Mo, 
                                 Covid19AI-Covid193Mo,    
                                 levels=design)
fit2 <- contrasts.fit(fit, contrast.matrix)
fit2 <- eBayes(fit2)
topTable(fit2) 
topTable(fit2,coef=1) 
topTable(fit2,coef=2) 
topTable(fit2,coef=3) 
# topTable(fit2,coef=4) try this!

# Store all of them
### covid19AI_vs_rest 
colnames(design) <- c("Intercept","Covid193Mo","Covid196Mo","Covid19AI")
contrast_covid19AI_vs_rest <- makeContrasts(Covid19AI-Covid193Mo, Covid19AI, 
                                 Covid19AI-Covid196Mo,    
                                 levels=design)
covid19AI_vs_rest <- contrasts.fit(vfit, contrast_covid19AI_vs_rest) 
covid19AI_vs_rest <- eBayes(covid19AI_vs_rest)

### covid196Mo_vs_rest 
contrast_covid196Mo_vs_rest <- makeContrasts(Covid196Mo-Covid193Mo, Covid196Mo, 
                                 Covid193Mo-Covid19AI,    
                                 levels=design)
covid196Mo_vs_rest <- contrasts.fit(vfit, contrast_covid196Mo_vs_rest) 
covid196Mo_vs_rest <- eBayes(covid19AI_vs_rest)

### Healthy_vs_rest 
contrast_Covid193Mo_vs_rest <- makeContrasts(Covid193Mo-Covid19AI, Covid193Mo, 
                                 Covid193Mo-Covid196Mo,    
                                 levels=design)
Covid193Mo_vs_rest <- contrasts.fit(vfit, contrast_Covid193Mo_vs_rest) 
Covid193Mo_vs_rest <- eBayes(Covid193Mo_vs_rest)


```

## ORA and Gene Set Enrichment analysis.

-   What do we need to do the analysis?
-   What are the tools required?
-   

```{r Prepare ORA and GSEA}

keytypes(org.Hs.eg.db)

# If we want to shift annotations:
ENSEMBL_vector <- mapIds(
  # Replace with annotation package for the organism relevant to your data
  org.Hs.eg.db,
  # The vector of gene identifiers we want to map
  keys = rownames(GSE198256_count),
  # Replace with the type of gene identifiers in your data
  keytype = "ENTREZID",
  # Replace with the type of gene identifiers you would like to map to
  column = "ENSEMBL",
  # In the case of 1:many mappings, return the
  # first one. This is default behavior!
  multiVals = "first"
)

# We would like a data frame we can join to the differential expression stats
gene_key_df <- data.frame(
  ensembl_id = ENSEMBL_vector,
  entrez_id = names(ENSEMBL_vector),
  stringsAsFactors = FALSE
) %>%
  # If an Ensembl gene identifier doesn't map to a gene symbol, drop that
  # from the data frame
  dplyr::filter(!is.na(ensembl_id))


```

## Lets conduct ORA.

-   What do we need to do the analysis?
-   What are the tools required?

```{r Run ORA}

# Step 1: determine genes of interest.
diff_table <- topTable(vfit,coef=1,p.value=0.01,number=10000) 
genes_dif<- rownames(diff_table )

# Step 2: determine background.
background_set <- unique(rownames(logCPM))

# Step 3: Determine gene sets.
msigdbr_species()
hs_msigdb_df <- msigdbr(species = "Homo sapiens")
head(hs_msigdb_df)

hs_kegg_df <- hs_msigdb_df %>%
  dplyr::filter(
    gs_cat == "C2", # This is to filter only to the C2 curated gene sets
    gs_subcat == "CP:KEGG" # This is because we only want KEGG pathways
  )

# Step 4: conduct ORA.

kegg_ora_results <- enricher(
  gene = genes_dif, # A vector of your genes of interest
  pvalueCutoff = 0.1, # Can choose a FDR cutoff
  pAdjustMethod = "BH", # Method to be used for multiple testing correction
  universe = background_set, # A vector containing your background set genes
  # The pathway information should be a data frame with a term name or
  # identifier and the gene identifiers
  TERM2GENE = dplyr::select(
    hs_kegg_df,
    gs_name,
    human_entrez_gene
  )
)

# Step 5: Visualize / explore
enrich_plot <- enrichplot::dotplot(kegg_ora_results)
enrich_plot
upset_plot <- enrichplot::upsetplot(kegg_ora_results)
upset_plot


# Step 6: EXERCISE: alternatives to KEGG?

# Now we want to look at top DE significant (p< 0.01) genes for the contrast covid19AI_vs_healthy
DEG_covid19AI_vs_Covid193Mo <- topTable(covid19AI_vs_rest,coef=1,p.value=0.01,number=10000) 
genes_DE <- rownames(DEG_covid19AI_vs_Covid193Mo)
background_set <- unique(rownames(logCPM)) # the background 
 # length(background_set)
# using GO:BP 
hs_go_df <- hs_msigdb_df %>%
  dplyr::filter(
    gs_subcat == "GO:BP" # This is because we only want KEGG pathways
  )

go_ora_results <- enricher(
  gene = genes_DE, # A vector of your genes of interest
  pvalueCutoff = 0.1, # Can choose a FDR cutoff
  pAdjustMethod = "BH", # Method to be used for multiple testing correction
  universe = background_set, # A vector containing your background set genes
  # The pathway information should be a data frame with a term name or
  # identifier and the gene identifiers
  TERM2GENE = dplyr::select(
    hs_go_df,
    gs_name,
    human_entrez_gene
  )
)

enrich_plot <- enrichplot::dotplot(go_ora_results)
enrich_plot

# note: see more examples: https://alexslemonade.github.io/refinebio-examples/03-rnaseq/pathway-analysis_rnaseq_01_ora.html#46_Over-representation_Analysis_(ORA) 

```

## Lets conduct GSEA.

```{r run GSEA}


# Step 1: determine genes of interest.
diff_table_all <- topTable(vfit,coef=1,p.value=1,number=nrow(logCPM)) 

# Step 2: determine background.

# Step 3: Determine gene sets.

msigdbr_species()
hs_msigdb_df <- msigdbr(species = "Homo sapiens")
head(hs_msigdb_df)

hs_kegg_df <- hs_msigdb_df %>%
  dplyr::filter(
    gs_cat == "C2", # This is to filter only to the C2 curated gene sets
    gs_subcat == "CP:KEGG" # This is because we only want KEGG pathways
  )


# Step 4: conduct GSEA
list_ordered <- diff_table_all[,"B"]
names(list_ordered) <- rownames(diff_table_all)
  
  
gsea_results <- GSEA(
  geneList = list_ordered, # Ordered ranked gene list
  minGSSize = 25, # Minimum gene set size
  maxGSSize = 500, # Maximum gene set set
  pvalueCutoff = 0.05, # p-value cutoff
  eps = 0, # Boundary for calculating the p value
  seed = TRUE, # Set seed to make results reproducible
  pAdjustMethod = "BH", # Benjamini-Hochberg correction
  TERM2GENE = dplyr::select(
    hs_kegg_df,
    gs_name,
    human_entrez_gene
  )
)



# Step 5: Visualize / explore
head(gsea_results@result)

gsea_result_df <- data.frame(gsea_results@result)
gsea_result_df %>%
  # This returns the 3 rows with the largest NES values
  dplyr::slice_max(NES, n = 3)

most_positive_nes_plot <- enrichplot::gseaplot(
  gsea_results,
  geneSetID = "KEGG_JAK_STAT_SIGNALING_PATHWAY",
  title = "KEGG_JAK_STAT_SIGNALING_PATHWAY",
  color.line = "#0d76ff"
)
most_positive_nes_plot

gsea_result_df %>%
  # Return the 3 rows with the smallest (most negative) NES values
  dplyr::slice_min(NES, n = 3)

most_negative_nes_plot <- enrichplot::gseaplot(
  gsea_results,
  geneSetID = "KEGG_SPLICEOSOME",
  title = "KEGG_SPLICEOSOME",
  color.line = "#0d76ff"
)
most_negative_nes_plot


# Step 6: EXERCISE: alternatives to KEGG?

# using GO:BP 
hs_go_df <- hs_msigdb_df %>%
  dplyr::filter(
    gs_subcat == "GO:BP" # This is because we only want KEGG pathways
  )

list_ordered <- diff_table_all[,"B"]
names(list_ordered) <- rownames(diff_table_all)
  
  
gsea_results_go <- GSEA(
  geneList = list_ordered, # Ordered ranked gene list
  minGSSize = 25, # Minimum gene set size
  maxGSSize = 500, # Maximum gene set set
  pvalueCutoff = 0.05, # p-value cutoff
  eps = 0, # Boundary for calculating the p value
  seed = TRUE, # Set seed to make results reproducible
  pAdjustMethod = "BH", # Benjamini-Hochberg correction
  TERM2GENE = dplyr::select(
    hs_go_df,
    gs_name,
    human_entrez_gene
  )
)

head(gsea_results_go@result)

# Step 7: EXERCISE: compare GSEA vs ORA?

ORA_res_go <- go_ora_results@result
GSEA_res_go <- gsea_results_go@result

common_gene_sets <- intersect(GSEA_res_go$ID, ORA_res_go$ID)
length(common_gene_sets)
nrow(GSEA_res_go)
nrow(ORA_res_go)

# comparing the pvalue of common_gene_sets
GSEA_res_go_sub <- GSEA_res_go[GSEA_res_go$ID %in% common_gene_sets, ]
 # nrow(GSEA_res_go_sub)
ORA_res_go_sub <- ORA_res_go[ORA_res_go$ID %in% common_gene_sets, ]
 # nrow(ORA_res_go_sub)
GSEA_res_go_sub <- GSEA_res_go_sub[order(GSEA_res_go_sub$ID), ]
ORA_res_go_sub <- ORA_res_go_sub[order(ORA_res_go_sub$ID), ]

correlation <- cor(-log10(GSEA_res_go_sub$p.adjust), ORA_res_go_sub$p.adjust)
plot(GSEA_res_go_sub$p.adjust, ORA_res_go_sub$p.adjust)


### 

```

## Lets conduct GeneSetCluster.

```{r prepare GSEA}
# Healthy vs Group Covid19 
# We prepare a function from the previous analysis

# Covid19AI vs Covid193Mo
Diff_AIvs3Mo <- topTable(fit2,coef=1,p.value=1,number=nrow(logCPM))
#Covid19AI vs Covid196Mo 
Diff_AIvs6Mo <- topTable(fit2,coef=3,p.value=1,number=nrow(logCPM))


hs_msigdb_df <- msigdbr(species = "Homo sapiens")
hs_kegg_df <- hs_msigdb_df %>%
  dplyr::filter(
    gs_cat == "C2", # This is to filter only to the C2 curated gene sets
    gs_subcat == "CP:KEGG" # This is because we only want KEGG pathways
  )


doGSEA <- function(diff_table) {
  list_ordered <- diff_table[,"B"]
  names(list_ordered) <- rownames(diff_table)
  
  return(GSEA(
    geneList = list_ordered, # Ordered ranked gene list
    minGSSize = 25, # Minimum gene set size
    maxGSSize = 500, # Maximum gene set set
    pvalueCutoff = 0.05, # p-value cutoff
    eps = 0, # Boundary for calculating the p value
    seed = TRUE, # Set seed to make results reproducible
    pAdjustMethod = "BH", # Benjamini-Hochberg correction
    TERM2GENE = dplyr::select(
      hs_kegg_df,
      gs_name,
      human_entrez_gene
    )
  ))
}

GSEA_AIvs3Mo <- doGSEA(Diff_AIvs3Mo)
GSEA_AIvs6Mo <- doGSEA(Diff_AIvs6Mo)

```
```{r install GeneSetCluster}
######## Check if the required packages are installed ########
packages <- c("limma","stats","methods","RColorBrewer","clusterProfiler","GGally",
              "network","clustree","readxl","org.Hs.eg.db",
              "org.Mm.eg.db","cluster","factoextra","STRINGdb","WebGestaltR","stringr",
              "AnnotationDbi","ComplexHeatmap","GO.db","GetoptLong","bigstatsr","colorRamp2",
              "cowplot","doParallel","dplyr","foreach","ggdendro","ggnewscale","ggplot2",
              "ggtree","ggwordcloud","grid","httr","jsonlite","parallel","patchwork","pbapply",
              "reshape2","rgl","seriation","simplifyEnrichment","slam","tidyverse","umap",
               "utils","grDevices")

new.packages <- packages[!(packages %in% installed.packages()[,"Package"])]

install.packages(new.packages)


```


```{r run GeneSetCluster}
install.packages('GeneSetCluster_1.99.tar.gz', repos=NULL, type='source')
library(GeneSetCluster)

# IMPORTANT when created manually, it is assumed that the pathways have been filtered by p-value and minimum number of genes per pathway
# Make sure you have filtered your data
GSEA.Object1Manual <- ObjectCreator(Pathways = c(GSEA_AIvs3Mo@result$ID, 
                                                 GSEA_AIvs6Mo@result$ID),
                                    Molecules = c(GSEA_AIvs3Mo@result$core_enrichment, 
                                                  GSEA_AIvs6Mo@result$core_enrichment),
                                    Groups = c(rep("GSEA_AIvs3Mo", times=nrow(GSEA_AIvs3Mo@result)), 
                                               rep("GSEA_AIvs6Mo", times=nrow(GSEA_AIvs6Mo@result))),
                                    Pvalues = c(GSEA_AIvs3Mo@result$p.adjust,  # optional
                                                GSEA_AIvs6Mo@result$p.adjust),
                                    enrichmentScore = c(GSEA_AIvs3Mo@result$NES, # optional
                                                        GSEA_AIvs6Mo@result$NES), 
                                    structure = "ENTREZID", Type = "", sep = "/",
                                    Source = "GSEA", organism = "org.Hs.eg.db")

GSEA.Object2 <- CombineGeneSets(Object = GSEA.Object1Manual,
                                combineMethod = "Standard", threads = 8)


OptimalGeneSets(Object = GSEA.Object2, 
                uniquePathway = FALSE, # consider all the pathways (also repeated) or the unique pathways
                method = "silhouette", max_cluster= 24, cluster_method = "kmeans", main= "Kmeans for 24 clusters")

OptimalGeneSets(Object = GSEA.Object2, 
                uniquePathway = TRUE, # consider all the pathways (also repeated) or the unique pathways
                method = "silhouette", max_cluster= 24, cluster_method = "kmeans", main= "Kmeans for 24 clusters")
# in both cases the optimal cluster is 2

GSEA.Object3 <- ClusterGeneSets(Object = GSEA.Object2, 
                                clusters = 2, # consider all the pathways (also repeated) or the unique pathways
                                method = "Hierarchical", # Hierarchical clustering or kmeans
                                order = "cluster",
                                molecular.signature = "All")

# plot results for both all pathways and unique pathways
plotnounique <- PlotGeneSets(GSEA.Object3, 
                             uniquePathways = FALSE, 
                             wordcloud = FALSE, # wordcloud only supported for GO terms
                             doORA = T) # do ora per cluster

plotunique <- PlotGeneSets(GSEA.Object3, 
                           uniquePathways = TRUE, 
                           wordcloud = FALSE, # wordcloud only supported for GO terms
                           doORA = T) # do ora per cluster


# let's say we are interested in exploring cluster 2 in plotunique. Lets break up this cluste for further analysis 

plotoptimalcluster2 <- OptimalGeneSets(Object = GSEA.Object3, 
                uniquePathway = TRUE, # consider all the pathways (also repeated) or the unique pathways
                cluster = 2, # which cluster
                method = "silhouette", max_cluster= 24, cluster_method = "kmeans", main= "Kmeans for 24 clusters in cluster 1")

plotoptimalcluster2 # optimal 2 break up cluster 2 in 2 clusters

GSEA.Object3breakup <- BreakUpCluster(GSEA.Object3, 
                                      breakup.cluster = 2, # which cluster
                                      sub.cluster = 2, # in how many cluster split up
                                      uniquePathways = TRUE) # conside unique pathways

plotuniquebreakup <- PlotGeneSets(GSEA.Object3breakup, 
                                  uniquePathways = TRUE, 
                                  wordcloud = FALSE, # wordcloud only supported for GO terms
                                  doORA = T) # do ora per cluster

plotuniquebreakup

# Now break up the cluster 1 
plotoptimalcluster1 <- OptimalGeneSets(Object = GSEA.Object3, 
                uniquePathway = TRUE, # consider all the pathways (also repeated) or the unique pathways
                cluster = 1, # which cluster
                method = "silhouette", max_cluster= 24, cluster_method = "kmeans", main= "Kmeans for 24 clusters in cluster 1")

plotoptimalcluster1 # optimal 1 break up cluster 1 in 9 clusters

GSEA.Object3breakup2 <- BreakUpCluster(GSEA.Object3breakup, 
                                      breakup.cluster = 1, # which cluster
                                      sub.cluster = 9, # in how many cluster split up
                                      uniquePathways = TRUE) # conside unique pathways

plotuniquebreakup2 <- PlotGeneSets(GSEA.Object3breakup2, 
                                   uniquePathways = TRUE, 
                                   wordcloud = FALSE, # wordcloud only supported for GO terms
                                   doORA = T) # do ora per cluster

plotuniquebreakup2


```